src/app/services/canvas.service.ts

```diff
@@ -0,0 +1,222 @@
+        this.tracingService.traceCreatePostClient();
+        this.fabricUtils._canvas.add(fabricPost);
+    }
+    
+    addPostServer(post: any): any {
+        this.tracingService.traceCreatePostServer(post.postID, post.title, post.desc).then(() => {
+            this.postsService.create(post);
+        })
+    }
+
+    modifyPostClient(post: Post, title: string, desc: string): any {
+        this.tracingService.traceModifyPostClient();
+
+        let obj: any = this.fabricUtils.getObjectFromId(post.postID);
+        // check if post is on board
+        if (obj){
+            obj = this.fabricUtils.updatePostTitleDesc(obj, title, desc)
+            obj.set({ title: title, desc: desc, canvasEvent: CanvasPostEvent.TITLE_CHANGE })
+            this.fabricUtils._canvas.renderAll()
+    
+            obj = this.fabricUtils.toJSON(obj)
+        }
+        // bucket only so fabricObject is {}
+        else{
+            obj ="{}"
+        }
+        
+        return obj;
+    }
+
+    modifyPostServer(obj: any, post: Post, title: string, desc: string): void {
+        this.postsService.update(post.postID, { fabricObject: obj, title: title, desc: desc }).then(() => {
+            this.tracingService.traceModifyPostServer(post.postID, title, desc);
+        });
+    }
+
+    createCommentClient(comment: any, comments: any): void {
+        this.tracingService.traceCreateCommentClient();
+        comments.push(comment);
+    }
+
+    createCommentServer(comment: any): void {
+        this.commentService.add(comment).then(() => {
+            this.tracingService.traceCreateCommentServer(comment.commentID, comment.comment);
+        });
+    }
+
+    likeModalPostClient(like: any, likes: any): void {
+        this.tracingService.traceVotedPostClient();
+        likes.push(like);
+    }
+
+    likeModalPostServer(postId: string, like: any): void {
+        this.likesService.add(like).then(() => {
+            this.tracingService.traceVotedPostServer(postId, 1);
+        });
+    }
+
+    async unlikeModalPostClient(likes: any, isLiked: any, userId: string): Promise<[any, any]> {
+        await this.tracingService.traceVotedPostClient();
+        isLiked = null;
+        likes = likes.filter(like => like.likerID != userId);
+        return [likes, isLiked];
+    }
+
+    unlikeModalPostServer(postId: string, likeId: string): void {
+        this.likesService.remove(likeId).then(() => {
+            this.tracingService.traceVotedPostServer(postId, 0);
+        });
+    }
+
+    likeCanvasPostClient(): void {
+        this.tracingService.traceVotedPostClient();
+    }
+    
+    likeCanvasPostServer(postId: string, userId: string, boardId: string): void {
+        this.likesService.add({
+            likeID: Date.now() + '-' + userId,
+            likerID: userId,
+            postID: postId,
+            boardID: boardId
+        }).then(() => this.tracingService.traceVotedPostServer(postId, 1));
+    }
```

Comment:
```suggestion
    likeCanvasPost(postId: string, userId: string, boardId: string): void {
        this.tracingService.traceVotedPostClient();
        this.likesService.add({
            likeID: Date.now() + '-' + userId,
            likerID: userId,
            postID: postId,
            boardID: boardId
        }).then(() => this.tracingService.traceVotedPostServer(postId, 1));
    }
```
Why can't we do a single function instead of having 2 separate ones? This would decrease some of the complexity.

Reply:
That's what I thought before too..but I think the separation has some advantages such as:

1. Clear separation between what's done on the client side, and what's done on the server side
2. If there's a bug, it's much easier to catch which one causes. For example, I was able to locate the "double-tracing" incident (which, as we discussed, was intentional for now) caused by deleting a post really quickly. It would have been slightly harder if they were combined into one

Reply:
I still think splitting it into 2 functions is excessive since we're kind've forcing the whole client vs server thing across the entire codebase. 

It should already be clear if we're calling 1 function then both client and server tracing will happen within that 1 function. Having a client and server function for each action just seems a little too much in my opinion.

Reply:
@Ammar-T how about the second point?

Reply:
I think we'll just sorta have to deal with it lol. You're right it would make it easier to debug, but not sure if duplicating all functions would be worth it.

Reply:
@l3n0ire what do you think about this?

Reply:
I think its better to keep it one function. Like Ammar mentioned, we'd have to make similar changes across the whole code base. You can debug it by putting a break point after the client code for ex.